module/zfs/vfs_ratelimit.c

```diff
@@ -0,0 +1,664 @@
+			if (slot->rts_counts[type] + count <=
+			    rl->rl_limits[type]) {
+				slot->rts_counts[type] += count;
+				break;
+			}
+
+			/*
+			 * This request is too big to fit into a single slot,
+			 * ie. a single request exceeds the limit or this and
+			 * the previous requests exceed the limit.
+			 */
+
+			/*
+			 * Fit as much as we can into the current slot.
+			 */
+			count -= rl->rl_limits[type] - slot->rts_counts[type];
+			slot->rts_counts[type] = rl->rl_limits[type];
+
+			/*
+			 * Take the next slot (if already exists isn't aware of
+			 * our type yet), take an expired slot from the tail of
+			 * the list or allocate a new slot.
+			 */
+			slot = list_prev(&rl->rl_list, slot);
+			if (slot != NULL) {
+				ASSERT((slot->rts_types & typebit) == 0);
+				ASSERT(slot->rts_timeslot == nexttimeslot);
+				ASSERT0(slot->rts_counts[type]);
+
+				slot->rts_types |= typebit;
+			} else {
+				slot = list_tail(&rl->rl_list);
+				if (slot->rts_timeslot < now) {
+					list_remove(&rl->rl_list, slot);
+				} else {
+					slot = kmem_alloc(sizeof (*slot),
+					    KM_SLEEP);
+				}
+				slot->rts_types = typebit;
+				slot->rts_timeslot = nexttimeslot;
+				memset(slot->rts_counts, 0,
+				    sizeof (slot->rts_counts));
+				list_insert_head(&rl->rl_list, slot);
+			}
+
+			nexttimeslot++;
+		}
+
+		if (timeslot < slot->rts_timeslot) {
+			timeslot = slot->rts_timeslot;
+		}
+	}
+
+	mutex_exit(&rl->rl_lock);
+
+	return (timeslot);
+}
+
+static void
+vfs_ratelimit(objset_t *os, int types, const uint64_t *counts)
+{
+	dsl_dir_t *dd;
+	hrtime_t now, timeslot;
+
+	now = gettimeslot();
+	timeslot = 0;
+
+	/*
+	 * Prevents configuration changes when we have requests in-flight.
+	 */
+	rrm_enter_read(&os->os_spa->spa_ratelimit_lock, FTAG);
+
+	for (dd = ratelimit_first(os, types); dd != NULL;
+	    dd = ratelimit_parent(dd, types)) {
+		hrtime_t ts;
+
+		ts = ratelimit_account(dd->dd_ratelimit, types, now, counts);
+		if (ts > timeslot) {
+			timeslot = ts;
+		}
+	}
+
+	rrm_exit(&os->os_spa->spa_ratelimit_lock, FTAG);
+
+	if (timeslot > now) {
+		/*
+		 * Too much traffic, slow it down.
+		 */
+		delay((hz / RATELIMIT_RESOLUTION) * (timeslot - now));
+	}
+}
+
+/*
+ * For every data read we charge:
+ * - bytes of read bandwidth
+ * - bytes of total bandwidth
+ * - (bytes - 1) / blocksize + 1 of read operations
+ * - (bytes - 1) / blocksize + 1 of total operations
+ */
+void
+vfs_ratelimit_data_read(objset_t *os, size_t blocksize, size_t bytes)
+{
+	uint64_t counts[ZFS_RATELIMIT_NTYPES];
+	unsigned int types;
+
+	if (bytes == 0) {
+		return;
+	}
+	if (blocksize == 0) {
+		blocksize = bytes;
+	}
+
+	types =  (1 << ZFS_RATELIMIT_BW_READ);
+	types |= (1 << ZFS_RATELIMIT_BW_TOTAL);
+	types |= (1 << ZFS_RATELIMIT_OP_READ);
+	types |= (1 << ZFS_RATELIMIT_OP_TOTAL);
+
+	memset(counts, 0, sizeof (counts));
+	counts[ZFS_RATELIMIT_BW_READ] = bytes;
+	counts[ZFS_RATELIMIT_BW_TOTAL] = bytes;
+	counts[ZFS_RATELIMIT_OP_READ] = (bytes - 1) / blocksize + 1;
```

Comment:
```suggestion
	counts[ZFS_RATELIMIT_OP_READ] = (bytes + blocksize - 1) / blocksize;
```

Reply:
Not sure why are you suggesting this:) It seems to be identical, but I'm more use to doing/seeing (val-1)/div+1.

Reply:
I also can say that I am more used to do it my way. ;)  And my way should properly handle case of val = 0, IMHO unlike yours.  Sure you check for `bytes == 0` above, but I still don't feel very comfortable about it.